metrics: density: memory footprint script

grahamwhaley opened the pull request:
Add a script that grabs as many useful memory footprint data items
as we can, user-space and kernel, so we can monitor and anlyse where
our memory goes.
Can be configured for:
 - different workloads
 - cut off points of no. containers, min ram avail, max mem consumed etc.

Fixes: #536

Signed-off-by: Graham whaley <graham.whaley@intel.com>

grahamwhaley commented:
@MarioCarrilloA @bwarden ptal
Although this script is not suitable for direct integration into the CI or our testing scripts, it is a useful tool for data capture for further analysis.
There are some more features we'll hopefully add in the future, and ideally we'll add some supplementary scripts to do some post process and stats on the captured data, but I think it will be useful to have this script available as-is right now.

MarioCarrilloA commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+PAYLOAD_ARGS="${PAYLOAD_ARGS:-tail -f /dev/null}"
+PAYLOAD_RUNTIME_ARGS="${PAYLOAD_RUNTIME_ARGS:-}"
+
+#  mysql is a more medium sized workload
```
I think is a great idea to test with different images size, maybe could be good to do this using an option in the script, something like `./footprint_data --large-imgsize`, instead to edit script

grahamwhaley replied:
My theory was that if somebody wanted to test with a different image or settings then they'd write a wrapper script and modify the ENV and then invoke the script. I was hoping to keep the script itself from becoming too complex or 'all things to all men'. I placed the mysql and elasticsearch commented out sections both to show what we have been running, and as examples of how to modify and set up the extra ENV settings.
If you are OK with that, can we leave the additions out of this script, and later we could add a 'run_multiple_images' script for instance?

MarioCarrilloA replied:
yeah, sounds good :)

MarioCarrilloA commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+}
+
+function go() {
+	for i in `seq 1 $MAX_NUM_CONTAINERS`; do
```
this line could use `$()` instead of backquotes for command substitution,  the newer form solves some inconsistencies, however in this case `$()` and backquotes will have the same effect

grahamwhaley replied:
Good spot - my old fingers defaulted to backticks - I would have gone with $() if I'd spotted it :-) New version coming...

MarioCarrilloA commented:
lgtm

grahamwhaley commented:
rebased - let's see how the CI looks. Note, Travis is an expected fail due to recent disablement.

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+# helper function to get USS of prcess in arg1
+function get_proc_uss() {
+	item=$(sudo smem -t -P "^$1" | tail -1 | awk '{print $4}')
+	item=$(($item * 1024))
```
Could be just `((item*=1024))`

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+# helper function to get PSS of prcess in arg1
+function get_proc_pss() {
+	item=$(sudo smem -t -P "^$1" | tail -1 | awk '{print $5}')
+	item=$(($item * 1024))
```
same here

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	shim=$(get_proc_uss $CC_SHIM_PATH)
+	qemu=$(get_proc_uss $CC_QEMU_PATH)
+
+	total=$(($proxy + $shim + $qemu))
```
Sigils inside `(( ))` are unnecessary

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	shim=$(get_proc_pss $CC_SHIM_PATH)
+	qemu=$(get_proc_pss $CC_QEMU_PATH)
+
+	total=$(($proxy + $shim + $qemu))
```
same here

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	fi
+
+	item=$(sudo smem -t | tail -1 | awk '{print $5}')
+	item=$(($item * 1024))
```
same here

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+
+	# userspace
+	item=$(sudo smem -w | head -5 | tail -1 | awk '{print $3}')
+	item=$(($item * 1024))
```
same here

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	# Grabbing slab total from meminfo is easier than doing the math
+	# on slabinfo
+	item=$(cat /proc/meminfo | fgrep "Slab:" | awk '{print $2}')
+	item=$(($item * 1024))
```
same here

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+
+	# Grabbing slab total from meminfo is easier than doing the math
+	# on slabinfo
+	item=$(cat /proc/meminfo | fgrep "Slab:" | awk '{print $2}')
```
Could eliminate `cat` via `fgrep "Slab:" /proc/meminfo`

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+
+function get_memfree() {
+	mem_free=$(sudo smem -w | head -6 | tail -1 | awk '{print $4}')
+	mem_free=$(($mem_free * 1024))
```
Same here

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	# avail memory, from 'free'
+	item=$(free -b | head -2 | tail -1 | awk '{print $7}')
+	echo -n ", $item" >> $DATAFILE
+	avail_incr=$(($base_mem_avail-$item))
```
Sigils

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	# free memory from smem
+	item=$(get_memfree)
+	echo -n ", $item" >> $DATAFILE
+	free_incr=$(($base_mem_free-$item))
```
Sigils

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+
+	# Anon pages
+	item=$(cat /proc/meminfo | fgrep "AnonPages:" | awk '{print $2}')
+	item=$(($item * 1024))
```
`*=`

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	echo -n ", $free_incr" >> $DATAFILE
+
+	# Anon pages
+	item=$(cat /proc/meminfo | fgrep "AnonPages:" | awk '{print $2}')
```
~~cat~~

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	echo -n ", $item" >> $DATAFILE
+
+	# Mapped pages
+	item=$(cat /proc/meminfo | fgrep "Mapped:" | awk '{print $2}')
```
~~cat~~

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+
+	# Mapped pages
+	item=$(cat /proc/meminfo | fgrep "Mapped:" | awk '{print $2}')
+	item=$(($item * 1024))
```
`*=`

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	echo -n ", $item" >> $DATAFILE
+
+	# Cached
+	item=$(cat /proc/meminfo | grep "^Cached:" | awk '{print $2}')
```
~~cat~~

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+
+	# Cached
+	item=$(cat /proc/meminfo | grep "^Cached:" | awk '{print $2}')
+	item=$(($item * 1024))
```
`*=`

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+function grab_stats() {
+	# If configured, dump the caches so we get a more stable
+	# view of what our static footprint really is
+	if (($DUMP_CACHES)) ; then
```
Did you mean `[[ $DUMP_CACHES ]]`?

grahamwhaley replied:
well, I explicitly put in the comments that it should be set to 0 or 1 :-) - but, I'll flip it to set or not...

bwarden replied:
I was actually unfamiliar with the `(( ))` operator (embarrassing, since it's right above `[[ ]]` in the manpage). I just saw weird behavior with one test case, so I wasn't sure it was doing the right thing:
    btwarden@btwarden-ws:~$ DUMP_CACHES="0" if (($DUMP_CACHES)); then echo true; else echo false; fi
    bash: syntax error near unexpected token `('

But these test cases worked as expected, so your code is probably best:
    btwarden@btwarden-ws:~$ export DUMP_CACHES=0
    btwarden@btwarden-ws:~$ if (($DUMP_CACHES)); then echo true; else echo false; fi
    false
    btwarden@btwarden-ws:~$ export DUMP_CACHES=1
    btwarden@btwarden-ws:~$ if (($DUMP_CACHES)); then echo true; else echo false; fi
    true
    btwarden@btwarden-ws:~$ unset DUMP_CACHES
    btwarden@btwarden-ws:~$ if (($DUMP_CACHES)); then echo true; else echo false; fi
    false
    btwarden@btwarden-ws:~$ export DUMP_CACHES=foo
    btwarden@btwarden-ws:~$ if (($DUMP_CACHES)); then echo true; else echo false; fi
    false

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+
+function check_limits() {
+	mem_free=$(get_memfree)
+	if (($mem_free <= $MIN_MEMORY_FREE)); then
```
Sigils

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+	fi
+
+	mem_consumed=$(($base_mem_avail-$mem_free))
+	if (($mem_consumed >= $MAX_MEMORY_CONSUMED)); then
```
Sigils

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+
+function go() {
+	for i in $(seq 1 $MAX_NUM_CONTAINERS); do
+		docker run -tid --runtime=$RUNTIME $PAYLOAD_RUNTIME_ARGS $PAYLOAD $PAYLOAD_ARGS
```
Any good chance of having these vars expand in unfortunate ways? If so, could rework as arrays.

grahamwhaley replied:
so far they've been OK, but yes, it can get a little tricky if trying to pass, say, args to a runtime that does a `bash -c xxx` for instance. OK as is for now I think - we can fix if we run into a complex case we cannot handle...

bwarden commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,354 @@
+		echo "" >> $DATAFILE
+
+		# check if we have hit one of our limits and need to wrap up the tests
+		if (($(check_limits))); then
```
Did you mean `[[ $(check_limits) ]]`?

grahamwhaley replied:
well, check_limits returns 0 or 1 .... I'm happy to change, but can you elaborate on the advantages for me?

grahamwhaley commented:
thanks @bwarden - yeah, my 'bash foo' is not too strong - undoubtedly I've used the wrong type of evaluate here and there - I'll have a sweeping clean tomorrow.

grahamwhaley commented:
All but one [[ change applied - thanks @MarioCarrilloA @bwarden

grahamwhaley commented:
@bwarden @jodh-intel @mcastelino (after a bit of a hiatus) script updated, docs added - rebased and added to the README as well. Added an example to the docs of how to invoke the script form an uber-script that sets up the env vars etc.

jodh-intel commented on metrics/density/footprint_data.md:
```diff
@@ -0,0 +1,104 @@
+# Footprint data script details
+
+The `footprint_data.sh` test within this directory can be configured with a number
```
Could you add a very brief overview of what the script does?

jodh-intel commented on metrics/density/footprint_data.md:
```diff
@@ -0,0 +1,104 @@
+payload is executed. The `RUNTIME` and `PAYLOAD` arguments directly affect the
+payload execution via this line in the script:
+
+`docker run -tid --runtime=$RUNTIME $PAYLOAD_RUNTIME_ARGS $PAYLOAD $PAYLOAD_ARGS`
```
Could you add a `$ ` prefix to show it's a command.

jodh-intel commented on metrics/density/footprint_data.md:
```diff
@@ -0,0 +1,104 @@
+
+| Variable | Function
+| -------- | --------
+| PAYLOAD | The docker image to run
```
Not necessary but if you backtick all these variables (and the measurement items in the table below), it looks cuter ;)

Less cute, but since some (?) of these variables seem to have default values, it might actually be better to add a `usage()` to the script and have it dump out this table, but also show the default values?

jodh-intel commented on metrics/density/footprint_data.md:
```diff
@@ -0,0 +1,104 @@
+| DOCKERD_PATH | The path to the Docker `dockerd` binary (for `smem` measurements)
+| DUMP_CACHES | A flag to note if the system caches should be dumped before capturing stats
+| DATAFILE | Can be set to over-ride the default CSV results filename
+| DATAFILE | Can be set to over-ride the default CSV results filename
```
Duplicated line.

jodh-intel commented on metrics/density/footprint_data.md:
```diff
@@ -0,0 +1,104 @@
+
+| Item | Description
+| ---- | -----------
+| number | The n'th container run during this test
```
1-based indexing?

grahamwhaley replied:
It is :-) Sort of obviously I thought - in the csv it is really just the row number...
I'll see if I can find a way to clarify that.

jodh-intel commented on metrics/density/footprint_data.md:
```diff
@@ -0,0 +1,104 @@
+| Variable | Function
+| -------- | --------
+| PAYLOAD | The docker image to run
+| PAYLOAD_ARGS | Any arguments passed into the docker image
```
Could you mention in the brief intro that this script only works for the docker container manager.

jodh-intel commented on metrics/density/footprint_data.md:
```diff
@@ -0,0 +1,104 @@
+## Output files
+The names of the CSV files generated by the test are dictated by some of the parameters
+the test is utilising. The default filename is defined by the expansion of:
+`${results-${RUNTIME}-${PAYLOAD}.csv}`
```
It might be a useful feature to have a script option (`-n` / `--dry-run` or something) that will just display the name of the file that would be generated?

jodh-intel commented on metrics/density/footprint_data.sh:
```diff
@@ -0,0 +1,359 @@
+PAYLOAD_RUNTIME_ARGS="${PAYLOAD_RUNTIME_ARGS:-}"
+
+#  mysql is a more medium sized workload
+#PAYLOAD="${PAYLOAD:-mysql}"
```
Commented out lines. If these are generally useful, how about creating a `data/` directory with a `mysql.cfg`, `elasticsearch.cfg`, *etc*` and have a CLI option that will simply `source` one of those data files if specified.

grahamwhaley replied:
I'm not sold on the data dir and breaking the examples out into separate files.
I can make the commented out examples into more obvious examples - and now we have the same data in the doc file, it feels better to make the information parallel/reflect that here - I'll change the format to more comment style

grahamwhaley commented:
@jodh-intel - addressed most the comments, rebased and re-pushed.

jodh-intel commented:
lgtm

grahamwhaley commented:
Dropping the WIP, I believe this is now done.
@MarioCarrilloA @bwarden - any last thoughts etc?
@iphutch pls review - thx.

iphutch commented on metrics/density/footprint_data.md:
```diff
@@ -0,0 +1,107 @@
+invocation script.
+
+## Variables
+The script action can be modified by setting a number of environment variables.
```
Please add a blank line after each heading (lines 13, 39, 65)

grahamwhaley commented:
Spaces fixed - but then my local rebase seems to have picked up an Azure commit for some reason ... once the CI has finished I'll rebase to master and push again to see if we can clean that up.

chavafg commented:
lgtm
